client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/HeaderDecoder.java

```diff
@@ -0,0 +1,439 @@
+         int port = buf.readUnsignedShort();
+         addresses[i] = InetSocketAddress.createUnresolved(host, port);
+      }
+      return addresses;
+   }
+
+   public void setCodec(Codec codec) {
+      assert channel.eventLoop().inEventLoop();
+      if (configuration.version() == ProtocolVersion.PROTOCOL_VERSION_AUTO) {
+         this.codec = codec;
+         channel.attr(OperationChannel.OPERATION_CHANNEL_ATTRIBUTE_KEY).get().setCodec(codec);
+      }
+   }
+
+   private void invokeEvent(byte[] listenerId, Object cacheEvent) {
+      try {
+         dispatcher.getClientListenerNotifier().invokeEvent(listenerId, cacheEvent);
+      } catch (Exception e) {
+         HOTROD.unexpectedErrorConsumingEvent(cacheEvent, e);
+      }
+   }
+
+   @Override
+   public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+      if (operation != null && !operation.isDone()) {
+         HotRodOperation<?> op = operation;
+         // we null out the operation, just in case the response handling generates another exception
+         operation = null;
+         dispatcher.handleResponse(op, receivedMessageId, ctx.channel(), null, cause);
+      } else {
+         if (log.isTraceEnabled()) {
+            TransportException transportException = log.errorFromUnknownOperation(ctx.channel(), cause, ctx.channel().remoteAddress());
+            log.tracef(transportException, "Requesting %s close due to exception", ctx.channel());
+         }
+         handleClosing(ctx, cause);
+         ctx.close();
+      }
+   }
+
+   @Override
+   public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+      TransportException transportException = log.connectionClosed(channel.remoteAddress(), channel.remoteAddress());
+      handleClosing(ctx, transportException);
+      super.channelInactive(ctx);
+   }
+
+   void failoverClientListeners() {
+      for (byte[] listenerId : listeners) {
+         dispatcher.getClientListenerNotifier().failoverClientListener(listenerId);
+      }
+   }
+
+   private void handleClosing(ChannelHandlerContext ctx, Throwable t) {
+      if (closing) {
+         return;
+      }
+      closing = true;
+      dispatcher.handleChannelFailure(ctx.channel(), t);
+      for (HotRodOperation<?> op : incomplete.values()) {
+         try {
+            dispatcher.handleResponse(op, -1, ctx.channel(), null, t);
+         } catch (Throwable innerT) {
+            HOTROD.errorf(t, "Failed to complete %s", op);
+         }
+      }
+      failoverClientListeners();
+      incomplete.clear();
+   }
+
+   public CompletableFuture<Void> allCompleteFuture() {
+      return CompletableFuture.allOf(incomplete.values().toArray(new CompletableFuture[0]));
+   }
+
+   @Override
+   public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+      if (evt instanceof IdleStateEvent) {
+         // If we have incomplete operations this channel is not idle!
+         if (!incomplete.isEmpty()) {
```

Comment:
Idle from the server side would be ignored?

Reply:
Good catch, I think this method can now just be removed. If the connection is write idle we now send the ping and if the server never sends a response (read idle) we close it which would cause the operations to be resubmitted to a new channel.